### DISCUSSION TITLE
potential intercommunication system

### DISCUSSION THREAD

**Discussion by ohitstom** [Delusoire/bespoke#11] [Ideas]
# Rundown
my main idea is that we could employ the use of a websocket server (or similar sort of system) that could enable:
- listening to packets from inside of spotify
- send packets with javascript to be run

This system would still require bespoke to be running in the background of the users computer 24/7 but personally i believe that this is a non-issue and most would prefer a small interface using ~1% of the cpu and ~15mb of ram if it means spicetify/bespoke will always be initialized. the only annoying part would be setting up the logic for this, but go surely has some sort of pre-existing architecture for os startup applications / services i would assume.

https://medium.com/@antoharyanto/make-simple-chat-application-using-golang-websocket-and-vanilla-js-f600e8020961
# Positives
### 1. auto re-initialization after spotify updates:
if `bespoke init` has previously been ran, and if spotify is open, but we are getting no websocket response - then its clear we need to trigger a re-initialize.

we can then trigger a reload / restart of spotify through a variety of methods:
- ~~simple os commands~~
- using spotifys remote debugging port
- etc

### 2.  ability to start a bespoke update from within spotify if required
this would allow us to trigger whatever update system for modules or major releases that may exist in the future from a modal inside of spotify at the click of a button for users

### 3.  potentially no longer needing a proxy for changes within bespoke
with a system like this or similar we wouldnt need to use your proxy server and could communicate purely on localhost for stuff such as config changes through marketplace

### ~~4. enable dev-tools from within spotify~~
~~can be added to the current debug menu, would be achieved similarly to 1.~~

# Notes
i would really like a discussion on this and not a cut and dry yes or no decision, or if you could fill me in if there was future plans to solve all of these that im not aware of since im not in the discord.

i think the main concern is `"auto re-initialization after spotify updates"` in the list of positives, i believe if you think this intercommunication system doesnt make sense that we could fix this by creating some sort of `bespoke auto` flag which (again) would run in the background (after being ran at startup) constantly checking if xpui.spa is re-added to spotify/Apps and if it is run `bespoke init` related logic - there would then be no need for a websocket system but i think intercommunication between bespoke and spotify would enable ALOT of stuff to be done outside of this issue.

as for security concerns, i believe as long as we are careful to limit what type of data can be accepted from the go side of things, the user *shouldnt* be able to break out of spotify and it *should* be a closed loop.

lastly i understand if you feel this feature would make more sense for a v4 in future but this seems like a really simple thing to setup (i dont think im underestimating websocket simplicity)

**Comment by Delusoire** [comment]
I have a few issues with this but I don't disagree with the idea.

> listening to packets from inside of spotify

Do you mean eval-ing js inside of Spotify? That's a nope.

> go surely has some sort of pre-existing architecture for os startup applications / services i would assume.

Go sure is a "simple" language, but it's not high-level. It doesn't manage these sorts of things for you. We'd have to write platform-specific code to handle every environment variation. It's feasible on Windows. macOS has the most annoying environment of all (We'll probably have to package bespoke as a .app with all its shenanigans). And Linux has a bunch of different init systems (systems, openrc, runit, ...) all of which handle startup services differently.

> simple os commands

That's another nope. I'm strongly against the use of shell scripting inside of Go.

> 2. ability to start a bespoke update from within spotify if required

You can already update modules from within Spotify with Marketplace. But updating Bespoke itself will probably never happen.

> 4. enable dev-tools from within spotify

This too can already be done without needing a background daemon.

> (i dont think im underestimating websocket simplicity)

No, but you're underestimating the amount of effort it takes to implement a cross-platform solution for auto-starting a process.

TLDR: automating bespoke updates & initialization would be an enormous QoL improvement but writing a solution that works for every platform is quasi-impossible.

**Comment by ohitstom** [comment]
> Do you mean eval-ing js inside of Spotify? That's a nope.

i believe v4 had planned on doing this with rust anyways? why is this a nope? the loop would be closed unless someone modifies the source for bespoke and redistributes it maliciously...  - like unless someone modifies the go source the stuff being eval'ed cant be modified by a bad actor (regardless this part wouldnt be needed for re-initializing which is the main thing i think this system could finally solve)

if eval is an absolute no, the go side can send off keywords which would then be compared to functions within spotify and ran - that way no eval?

> No, but you're underestimating the amount of effort it takes to implement a cross-platform solution for auto-starting a process.

may i suggest we use the install scripts that will be made in future to add the required services/startup related stuff and that way the go code would just have to handle the websocket? i think that would be a good compromise - because bash/shell/ps in that department should be easy

> But updating Bespoke itself will probably never happen.

it could if this system were to be implemented 😉 

im more than willing to help with the install script side of things, and possibly the go - but i would need you to handle the js side of things (or atleast polish it up haha) 👍🏻 

(of course this sort of thing would be implemented *last* ideally, once everything else is stable)

**Comment by Delusoire** [comment]
> i believe v4 had planned on doing this with rust anyways?

What was planned for v4 is the custom uri scheme which is already implemented.

> if eval is an absolute no, the go side can send off keywords ...

The Go side doesn't have to send anything. What are you trying to achieve here exactly?

> may i suggest we use the install scripts that will be made in future to add the required services/startup related stuff and that way the go code would just have to handle the websocket?

That would work very well

> it could if this system were to be implemented 😉

I mean it doesn't seem like a good idea to me. Bespoke is a mod for Spotify, not the other way around. Turning Spotify into a control and command center for bespoke isn't appealing.

> (of course this sort of thing would be implemented last ideally, once everything else is stable)

Lately, it seems like bespoke is more stable than Spicetify v2, just saying ... 😄

Here's my solution to this problem:

Implement a new `bespoke daemon [enable|disable]` command, enable will enable the daemon and start it, disable will disable the daemon and stop it. The state of the daemon will be saved to the config file, so that on startup we only have to run `bespoke daemon` and the daemon figures if it's enabled or not. when running, the daemon will basically run `bespoke init` whenever it detects that our files are deleted from the Apps/ directory. The daemon (if enabled) will also start a WS server which will be probed once Marketplace is launched, if the server responds then we can use the WS connection, otherwise fallback to the custom URI scheme.

**Reply by ohitstom** [reply]
your solution encompasses what I was trying to explain (mb), sounds great!

anything I can help with?

**Comment by Delusoire** [comment]
> anything I can help with?

Uhmm sure, you can write the code for the ws server using [this](https://github.com/gorilla/websocket/blob/main/examples/echo/server.go) as a reference.

**Comment by Delusoire** [comment]
The last few commits added a very simple daemon and WebSocket server. The implementation is rudimentary and probably error-prone. So long-term testing is needed.
